src/service/grpc.cpp

```diff
@@ -40,6 +41,57 @@ grpc::ServerUnaryReactor *Grpc::CreateCollection(
 	return reactor;
 }
 
+grpc::ServerUnaryReactor *Grpc::RetrieveCollection(
+	grpc::CallbackServerContext *context, const gk::v1::RetrieveCollectionRequest *request,
+	gk::v1::Collection *response) {
+	auto *reactor = context->DefaultReactor();
+
+	try {
+		auto collection = datastore::RetrieveCollection(request->id());
+		map(collection, response);
+	} catch (const err::DatastoreCollectionNotFound &) {
+		reactor->Finish(grpc::Status(grpc::StatusCode::NOT_FOUND, "Document not found"));
+		return reactor;
+	} catch (...) {
+		reactor->Finish(grpc::Status(grpc::StatusCode::UNAVAILABLE, "Failed to retrieve data"));
+		return reactor;
+	}
+
+	reactor->Finish(grpc::Status::OK);
+	return reactor;
+}
+
+grpc::ServerUnaryReactor *Grpc::UpdateCollection(
+	grpc::CallbackServerContext *context, const gk::v1::UpdateCollectionRequest *request,
+	gk::v1::Collection *response) {
+	auto *reactor = context->DefaultReactor();
+
+	try {
+		auto collection = datastore::RetrieveCollection(request->id());
+		bool hasUpdates = false;
+		if (request->has_name()) {
+			collection.name(request->name());
+			hasUpdates = true;
+		}
+
+		if (hasUpdates) {
+			collection.store();
+		}
+
+		map(collection, response);
+	} catch (const err::DatastoreCollectionNotFound &) {
+		reactor->Finish(grpc::Status(grpc::StatusCode::NOT_FOUND, "Document not found"));
+		return reactor;
+	} catch (...) {
+		reactor->Finish(grpc::Status(grpc::StatusCode::UNAVAILABLE, "Failed to retrieve data"));
+		return reactor;
+	}
```

Comment:
🤔 I see why this is in one `try {} catch {}` block but it'll be unhelpful to return _Failed to retrieve data_ if `collection.store()` fails. 

At least we should update the error message so it's not misleading.

Ideally we should break this block into smaller blocks. Maybe [`std::optional`](https://en.cppreference.com/w/cpp/utility/optional) would be useful. 

e.g. 
```cpp
std::optional<datastore::Collection> collection = std::nullopt;

try {
  *collection = datastore::RetrieveCollection(request->id()); 
} catch (...) {
  // Handle read errors and return with error messages
}

// If we get here, we should have a collection read from db
collection->name(request->name());
collection->store();
```